File: src/strings/GroupAnagrams_49.java
package strings;

import java.util.*;

/**
 * Given an array of strings, group anagrams together.
 */
public class GroupAnagrams_49 {
    public static void main(String[] args) {
        System.out.println(printDim(new GroupAnagrams_49().groupAnagrams(new String[]{"boo", "bob"})));
        System.out.println(printDim(new GroupAnagrams_49().groupAnagrams(new String[]{"eat", "tea", "tan", "ate", "nat", "bat"})));
        System.out.println(printDim(new GroupAnagrams_49().groupAnagrams(new String[]{"cab", "tin", "pew", "duh", "may", "ill", "buy", "bar", "max", "doc"})));
        System.out.println(printDim(new GroupAnagrams_49().groupAnagrams(new String[]{"hos", "boo", "nay", "deb", "wow", "bop", "bob", "brr", "hey", "rye", "eve", "elf",
                "pup", "bum", "iva", "lyx", "yap", "ugh", "hem", "rod", "aha", "nam", "gap", "yea", "doc", "pen", "job", "dis", "max", "oho", "jed", "lye", "ram", "pup", "qua",
                "ugh", "mir", "nap", "deb", "hog", "let", "gym", "bye", "lon", "aft", "eel", "sol", "jab"})));
    }

    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();

        for (String str : strs) {
            String hash;
            int[] hashArray = new int[26];
            for (int i = 0; i < str.length(); i++) {
                hashArray[str.charAt(i) - 'a']++;
            }
            hash = Arrays.toString(hashArray);

            List<String> anagrams = map.getOrDefault(hash, new LinkedList<>());
            anagrams.add(str);
            map.put(hash, anagrams);
        }

        return new LinkedList<>(map.values());
    }

    public static String printDim(List<List<String>> list) {
        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < list.size(); i++) {
            builder.append(Arrays.toString(list.get(i).toArray()));
            builder.append("\n");
        }
        return builder.toString();
    }
}

File: src/hash_table/group_anagrams_49.py
from typing import List


# 0. nklogk - sorting, nk space
# 1. nk - hash
# 2. nk - identity by 26x vector n*k time, n*k space

def get_vector(str):
    vector = [0] * 26

    for c in str:
        index = ord(c) - ord('a')
        vector[index] += 1

    return tuple(vector)


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        result = []
        dict = {}  # vector: [strings] # abc, bca

        for str in strs:
            vector = get_vector(str)
            value = dict.get(vector, [])
            value.append(str)
            dict[vector] = value

        for arr in dict.values():
            result.append(arr)

        return result

print(Solution().groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]))
